service/dap/server.go

```diff
@@ -641,37 +651,32 @@ func isValidLaunchMode(launchMode interface{}) bool {
 func (s *Server) onDisconnectRequest(request *dap.DisconnectRequest) {
 	s.send(&dap.DisconnectResponse{Response: *newResponse(request.Request)})
 	if s.debugger != nil {
-		_, err := s.debugger.Command(&api.DebuggerCommand{Name: api.Halt}, nil)
-		if err != nil {
-			switch err.(type) {
-			case *proc.ErrProcessExited:
-				s.log.Debug(err)
-			default:
-				s.log.Error(err)
-			}
-		}
-		// We always kill launched programs
-		kill := s.config.Debugger.AttachPid == 0
+		// We always kill launched programs.
 		// In case of attach, we leave the program
-		// running but default, which can be
+		// running by default, which can be
 		// overridden by an explicit request to terminate.
-		if request.Arguments.TerminateDebuggee {
-			kill = true
-		}
-		err = s.debugger.Detach(kill)
-		if err != nil {
-			switch err.(type) {
-			case *proc.ErrProcessExited:
-				s.log.Debug(err)
-			default:
-				s.log.Error(err)
-			}
-		}
+		killProcess := s.config.Debugger.AttachPid == 0 || request.Arguments.TerminateDebuggee
+		s.stopDebugSession(killProcess)
 	} else {
 		s.stopNoDebugProcess()
 	}
 	// TODO(polina): make thread-safe when handlers become asynchronous.
-	s.signalDisconnect()
+	s.triggerServerStop()
+}
+
+func (s *Server) stopDebugSession(killProcess bool) {
+	_, err := s.debugger.Command(&api.DebuggerCommand{Name: api.Halt}, nil)
```

Comment:
Why doesn't rpc server halt in stop()?
https://github.com/go-delve/delve/blob/c223ef656da3ddbab3a58ea87108d4217731a01c/service/rpccommon/server.go#L91-L97? 
On a related note, the terminology collision here is unfortunate: https://github.com/go-delve/delve/blob/master/Documentation/api/ClientHowto.md#gracefully-ending-the-debug-session
From the RPC client point of view, it appears that
- Detach = detach debugger from the process (maybe kill it) + close client connection
- Disconnect = close client connection
But the client how-to uses a phrase "disconnecting a running program" as a case where Detach is called. And then we have DAP where disconnect request is sent to "disconnect from the debuggee and to terminate the debug adapter". Sigh.

Reply:
Wild guess - isn't it because it's the rpc client who is supposed to call Halt (or Disconnect) command  before the server stops. Not sure about the case where the debugging stops because client unexpectedly crashes or gets disconnected.

Reply:
That's precisely the case I was thinking about. The client takes care of Halt before Detach. But if there is an error reading rpc messages from client, the server will go into shutdown mode and call Stop() without halting.

Reply:
I don't think there's a particular reason, halting would probably be better. And the client howto should say detach where it says ""disconnecting a running program".